### ISSUE TITLE
API request, ListSinceBlock for zaddrs

### ISSUE THREAD

**Issue by BitHighlander** [zcash/zcash#1333]
In order to integrate Zcash and allow deposits in private payments for customers; exchanges and power users will need to monitor a large amount of addresses. In order to facilitate this it is extremely hard to look up on payments single addresses via api. As for a large exchange this could mean 100,000 api calls + for just looking for payments, and this needs to be continual. 

This is why we would like to see a listsinceblock that shows all transactions to every single address in the wallet.

Also, an alternative that is just as useful is supporting the walletnotify service.    This allows a push anytime a payment is received as well.

**Comment by bitcartel** [comment]
With bitcoind, what JSON-RPC calls do you currently make use of to achieve the above?

**Comment by nathan-at-least** [comment]
We need clarification about what @BitHighlander proposes a request should return.

There is already the `listsinceblock` call (see [Bitcoin API_calls_list wiki page](https://en.bitcoin.it/wiki/Original_Bitcoin_client/API_calls_list)) which returns the public data associated with all transactions involving any t-addr or z-addr known to the wallet.

These results don't show private data, including:

* which z-addrs are involved,
* the private memo plaintext,
* private value transfers.

Is this sufficient information for your uses, or do you need more information to be provided?

**Comment by nathan-at-least** [comment]
Note: We're focused mainly on high priority security fixes up to launch, not on feature improvements or user requests. However, having good support for exchanges is obviously a priority, so if we determine there's a specific request thats low-complexity to implement and likely to be widely useful, we may try to get it into this beta2.

**Comment by nathan-at-least** [comment]
We've decided to postpone this until after launch, since our focus for the Beta 2 and release candidate(s) is crucial bug fixes and security issues.

**Comment by daira** [comment]
It seems like a combination of listsinceblock and a new call to decrypt the contents of a given transaction, would be most useful.

**Comment by nathan-at-least** [comment]
This was requested by a user, so I'm bumping up it's priority. I've placed it in 1.0.6.

**Comment by nathan-at-least** [comment]
I assigned @str4d to ensure this doesn't get dropped. @str4d: If you can't get to this, can you delegate to someone else?

**Comment by str4d** [comment]
From a cursory look at `listsinceblock`, it looks like that should already be showing transactions involving JoinSplits - it just doesn't output any of the JoinSplit data (in particular the info that we can decrypt inside recipient notes). So we should just be able to extend that RPC, rather than needing to create a new one.

**Comment by str4d** [comment]
In fact, I see that @nathan-at-least already made this point in https://github.com/zcash/zcash/issues/1333#issuecomment-246864804, so what we actually need to decide is whether we want this RPC to show private data, or whether we should have a separate RPC for that.

**Comment by nathan-at-least** [comment]
Let's add a `z_listblock` which is identical to `listblock` except there's a new `includeDecryptions` `bool` parameter (or some better name) that defaults to `false`. If `false`, the result is the same as `listblock`. If `true`, it includes decrypted secrets.

**Comment by daira** [comment]
I don't really see the need for an ``includeDecryptions`` parameter, if the ``false`` case is exactly the same as ``listsinceblock``. We don't need to deprecate all of the Bitcoin RPC calls just for the sake of it. We can call the new RPC ``z_decryptsinceblock`` to make clear what it does.

**Comment by daira** [comment]
I think using ZMQ for notifications (#2020) might be preferable to [Bitcoin's walletnotify](https://bitcointalk.org/index.php?topic=203438.0) for both simplicity and performance.

**Comment by bitcartel** [comment]
@daira Digression: the notification layer should ideally be abstracted out so that the user can configure what service to use and the code is not tightly coupled to ZMQ e.g. new txid in mempool, fires notification, notification event handler picks up, and checks if it should pass onto zmq, twitter, logging service based on config.

**Comment by nathan-at-least** [comment]
This is already in-progress, but it's blocking on a few RPC changes we'll be discussing in our topical meeting tomorrow.

**Comment by daira** [comment]
@str4d: did we discuss everything needed to unblock you?

**Comment by nathan-at-least** [comment]
Bumped, see associated PR: https://github.com/zcash/zcash/pull/2105#issuecomment-300591110

**Comment by Tomas-M** [comment]
Is there any progress on this?

**Comment by daira** [comment]
This seems to have been forgotten; last progress on the PR was this comment: https://github.com/zcash/zcash/pull/2105#issuecomment-287978718

**Comment by str4d** [comment]
Now that we have `z_viewtransaction` (#4146), we should revisit #2105 and see whether it still makes sense or needs to be rewritten.

**Comment by str4d** [comment]
#2105 definitely needs to be rewritten (so we have closed it as obsolete), and I'm now of the opinion that we shouldn't try to maintain close compatibility to the existing `listsinceblock` RPC method.
